modules/touch/src/touchinteraction.cpp

```diff
@@ -678,100 +843,379 @@ void TouchInteraction::findSelectedNode(const std::vector<TuioCursor>& list) {
+        nullptr,
+        std::numeric_limits<double>::max()
+    };
+    
+
+    for (const TouchInputs& c : list) {
+        double xCo = 2 * (c.getLatestInput().x - 0.5);
+        double yCo = -2 * (c.getLatestInput().y - 0.5); // normalized -1 to 1 coordinates on screen
+        glm::dvec3 cursorInWorldSpace = camToWorldSpace *
+            glm::dvec3(glm::inverse(_camera->projectionMatrix()) *
+            glm::dvec4(xCo, yCo, -1.0, 1.0));
+        glm::dvec3 raytrace = glm::normalize(cursorInWorldSpace);
+
+        size_t id = c.getFingerId();
+
+        for (SceneGraphNode* node : selectableNodes) {
+            double boundingSphereSquared = static_cast<double>(node->boundingSphere()) *
+                                           static_cast<double>(node->boundingSphere());
+            glm::dvec3 camToSelectable = node->worldPosition() - camPos;
+            double intersectionDist = 0.0;
+            bool intersected = glm::intersectRaySphere(
+                camPos,
+                raytrace,
+                node->worldPosition(),
+                boundingSphereSquared,
+                intersectionDist
+            );
+            if (intersected) {
+                glm::dvec3 intersectionPos = camPos + raytrace * intersectionDist;
+                glm::dvec3 pointInModelView = glm::inverse(node->worldRotationMatrix()) *
+                                              (intersectionPos - node->worldPosition());
+
+                // Add id, node and surface coordinates to the selected list
+                auto oldNode = std::find_if(
+                    newSelected.begin(),
+                    newSelected.end(),
+                    [id](const DirectInputSolver::SelectedBody& s) { return s.id == id; }
+                );
+                if (oldNode != newSelected.end()) {
+                    double oldNodeDist = glm::length(
+                        oldNode->node->worldPosition() - camPos
+                    );
+                    if (glm::length(camToSelectable) < oldNodeDist) {
+                         // new node is closer, remove added node and add the new one
+                         // instead
+                        newSelected.pop_back();
+                        newSelected.push_back({ id, node, pointInModelView });
+                    }
+                }
+                else {
+                    newSelected.push_back({ id, node, pointInModelView });
+                }
+            }
+
+            // Compute locations in view space to perform the picking
+            glm::dvec4 clip = glm::dmat4(_camera->projectionMatrix()) *
+                              _camera->combinedViewMatrix() *
+                              glm::vec4(node->worldPosition(), 1.0);
+            glm::dvec2 ndc = clip / clip.w;
+
+            // If the object is not in the screen, we dont want to consider it at all
+            if (ndc.x >= -1.0 && ndc.x <= 1.0 && ndc.y >= -1.0 && ndc.y <= 1.0) {
```

Comment:
For these types of complex if statements, I prefer something like this to make it more readable (and prevents need for a comment):
```
bool isVisibleX = (ndc.x >= -1.0 && ndc.x <= 1.0);
bool isVisibleY = (ndc.y >= -1.0 && ndc.y <= 1.0);
if (isVisibleX && isVisibleY) {
```

Reply:
(not my code since it's part of a refactor from tuio-points)
I agree with you and will change